# Online Retail Sales Analysis

## 1. Project Overview

### 1.1 Introduction

This project focuses on analyzing an online retail transaction dataset using Python. The purpose is to explore retail sales data, identify patterns, understand product performance, and extract useful business insights.

### 1.2 Objectives

* Explore and understand the retail transaction dataset.
* Investigate data quality and prepare the data for analysis.
* Calculate key sales performance metrics.
* Analyze product sales and revenue trends.
* Visualize findings and communicate actionable business insights.

### 1.3 Tools and Technologies

* **Python:** Primary programming language.
* **Pandas:** Data manipulation and analysis.
* **NumPy:** Numerical operations.
* **Matplotlib:** Data visualization.
* **Seaborn:** Statistical visualization.
* **Jupyter Notebook:** Interactive analysis environment.
* **Git and GitHub:** Version control and project publication.

### 1.4 Project Scope

The analysis covers transaction-level retail data, including products, quantities, unit prices, invoice dates, and countries. The project will focus on sales performance, data quality, product analysis, and time-based trends.

## 2. Dataset Description

### 2.1 Data Source

The dataset used in this project is the **Online Retail** dataset from the UCI Machine Learning Repository.

* **Dataset:** Online Retail
* **Source:** UCI Machine Learning Repository
* **Period:** December 2010 – December 2011
* **File format:** Excel (`.xlsx`)
* **Original size:** 541,909 transaction records
* **Number of attributes:** 8

### 2.2 Dataset Overview

The dataset contains transactional records from a UK-based online retail business. Each record represents a product line within an invoice.

### 2.3 Data Dictionary

| Column      | Description                                                                  |
| ----------- | ---------------------------------------------------------------------------- |
| InvoiceNo   | Unique invoice identifier; invoices beginning with C indicate cancellations. |
| StockCode   | Product identifier.                                                          |
| Description | Product name or description.                                                 |
| Quantity    | Number of units purchased or returned.                                       |
| InvoiceDate | Date and time of the transaction.                                            |
| UnitPrice   | Price per unit in GBP (£).                                                   |
| CustomerID  | Customer identifier, where available.                                        |
| Country     | Country associated with the transaction.                                     |

### 2.4 Analytical Scope

The dataset will be used to investigate sales revenue, product performance, monthly sales trends, geographic performance, and transaction data quality.

The dataset does not contain predefined product categories, profit margins, or geographic regions. These will not be assumed in the analysis.


## 3. Import Libraries and Load Data

### 3.1 Import Libraries

The required Python libraries are imported to support data manipulation, numerical calculations, and visualization.

### 3.2 Load the Dataset

The original Online Retail Excel file is loaded into a Pandas DataFrame. The original dataset is preserved so that any future cleaning operations can be performed on a separate copy.


In [43]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Display settings
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 20)
pd.set_option("display.float_format", "{:,.2f}".format)

print("Libraries imported successfully!")

Libraries imported successfully!


In [44]:
# Load the original dataset

df = pd.read_excel("../data/Online Retail.xlsx")

print("Dataset loaded successfully!")

Dataset loaded successfully!


## 4. Initial Data Inspection

### 4.1 Introduction

Initial data inspection helps understand the structure, size, and characteristics of the dataset before performing data cleaning and analysis.

The following checks will be performed:

* Display the first 10 records.
* Identify the number of rows and columns.
* Examine column names.
* Review the data types of each column.
* Generate basic descriptive statistics.

These checks provide an overview of the dataset and help identify areas that require further investigation.


In [45]:
# Display the first 10 records
df.head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,"17,850.00",United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,"17,850.00",United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,"17,850.00",United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,"17,850.00",United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,"17,850.00",United Kingdom
5,536365,22752,SET 7 BABUSHKA NESTING BOXES,2,2010-12-01 08:26:00,7.65,"17,850.00",United Kingdom
6,536365,21730,GLASS STAR FROSTED T-LIGHT HOLDER,6,2010-12-01 08:26:00,4.25,"17,850.00",United Kingdom
7,536366,22633,HAND WARMER UNION JACK,6,2010-12-01 08:28:00,1.85,"17,850.00",United Kingdom
8,536366,22632,HAND WARMER RED POLKA DOT,6,2010-12-01 08:28:00,1.85,"17,850.00",United Kingdom
9,536367,84879,ASSORTED COLOUR BIRD ORNAMENT,32,2010-12-01 08:34:00,1.69,"13,047.00",United Kingdom


In [46]:
# Number of rows and columns
print("Dataset dimensions:", df.shape)
print("Total rows:", df.shape[0])
print("Total columns:", df.shape[1])

Dataset dimensions: (541909, 8)
Total rows: 541909
Total columns: 8


In [47]:
# Display all column names
print("Column names:")
print(df.columns.tolist())

Column names:
['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'UnitPrice', 'CustomerID', 'Country']


In [48]:
# Examine the data types
print(df.dtypes)

InvoiceNo              object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[us]
UnitPrice             float64
CustomerID            float64
Country                   str
dtype: object


In [49]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[us]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  str           
dtypes: datetime64[us](1), float64(2), int64(1), object(3), str(1)
memory usage: 33.1+ MB


In [50]:
# Summary statistics for numerical columns
display(df.describe())

,Quantity,InvoiceDate,UnitPrice,CustomerID
count,"541,909.00",541909,"541,909.00","406,829.00"
mean,9.55,2011-07-04 13:34:57.156386,4.61,"15,287.69"
min,"-80,995.00",2010-12-01 08:26:00,"-11,062.06","12,346.00"
25%,1.00,2011-03-28 11:34:00,1.25,"13,953.00"
50%,3.00,2011-07-19 17:17:00,2.08,"15,152.00"
75%,10.00,2011-10-19 11:27:00,4.13,"16,791.00"
max,"80,995.00",2011-12-09 12:50:00,"38,970.00","18,287.00"
std,218.08,NaN,96.76,"1,713.60"


### 4.2 Initial Observations

Based on the initial inspection:

1. The dataset contains 541,909 rows and 8 columns.
2. The dataset contains transaction identifiers, product information, quantities, prices, dates, customer IDs, and countries.
3. The `InvoiceDate` column is already stored in a datetime format.
4. The `Quantity` column is an integer, while `UnitPrice` is a floating-point number.
5. The dataset includes both numerical and categorical attributes.

These observations provide an initial understanding of the dataset. Missing values, duplicates, and unusual transactions will be investigated separately during the data-quality investigation stage.


## 5. Data Quality Investigation

### 5.1 Introduction

Data quality investigation is an important step in preparing data for reliable analysis. The purpose of this stage is to identify missing values, duplicate records, invalid dates, unusual quantities, incorrect prices, and cancelled transactions.

The investigation will help determine which records require cleaning and which unusual values represent legitimate business transactions.

The following areas will be examined:

* Missing values
* Duplicate records
* Unusual quantities
* Invalid or unusual prices
* Missing or invalid dates
* Cancelled transactions

The original dataset will remain unchanged during this investigation.


In [51]:
# Count missing values in each column
missing_count = df.isnull().sum()

# Calculate missing percentages
missing_percentage = (
    df.isnull().mean() * 100
).round(2)

# Create a summary table
missing_summary = pd.DataFrame({
    "Missing Count": missing_count,
    "Missing Percentage": missing_percentage
})

display(
    missing_summary.sort_values(
        "Missing Percentage",
        ascending=False
    )
)

,Missing Count,Missing Percentage
CustomerID,135080,24.93
Description,1454,0.27
StockCode,0,0.00
InvoiceNo,0,0.00
Quantity,0,0.00
InvoiceDate,0,0.00
UnitPrice,0,0.00
Country,0,0.00


In [52]:
# Examine records with missing customer IDs
missing_customers = df[df["CustomerID"].isna()]

print("Rows with missing CustomerID:",
      len(missing_customers))

display(missing_customers.head(10))

Rows with missing CustomerID: 135080


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.00,NaN,United Kingdom
1443,536544,21773,DECORATIVE ROSE BATHROOM BOTTLE,1,2010-12-01 14:32:00,2.51,NaN,United Kingdom
1444,536544,21774,DECORATIVE CATS BATHROOM BOTTLE,2,2010-12-01 14:32:00,2.51,NaN,United Kingdom
1445,536544,21786,POLKADOT RAIN HAT,4,2010-12-01 14:32:00,0.85,NaN,United Kingdom
1446,536544,21787,RAIN PONCHO RETROSPOT,2,2010-12-01 14:32:00,1.66,NaN,United Kingdom
1447,536544,21790,VINTAGE SNAP CARDS,9,2010-12-01 14:32:00,1.66,NaN,United Kingdom
1448,536544,21791,VINTAGE HEADS AND TAILS CARD GAME,2,2010-12-01 14:32:00,2.51,NaN,United Kingdom
1449,536544,21801,CHRISTMAS TREE DECORATION WITH BELL,10,2010-12-01 14:32:00,0.43,NaN,United Kingdom
1450,536544,21802,CHRISTMAS TREE HEART DECORATION,9,2010-12-01 14:32:00,0.43,NaN,United Kingdom
1451,536544,21803,CHRISTMAS TREE STAR DECORATION,11,2010-12-01 14:32:00,0.43,NaN,United Kingdom


In [53]:
missing_descriptions = df[
    df["Description"].isna()
]

print("Missing descriptions:",
      len(missing_descriptions))

display(missing_descriptions.head(10))

Missing descriptions: 1454


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.00,NaN,United Kingdom
1970,536545,21134,NaN,1,2010-12-01 14:32:00,0.00,NaN,United Kingdom
1971,536546,22145,NaN,1,2010-12-01 14:33:00,0.00,NaN,United Kingdom
1972,536547,37509,NaN,1,2010-12-01 14:33:00,0.00,NaN,United Kingdom
1987,536549,85226A,NaN,1,2010-12-01 14:34:00,0.00,NaN,United Kingdom
1988,536550,85044,NaN,1,2010-12-01 14:34:00,0.00,NaN,United Kingdom
2024,536552,20950,NaN,1,2010-12-01 14:34:00,0.00,NaN,United Kingdom
2025,536553,37461,NaN,3,2010-12-01 14:35:00,0.00,NaN,United Kingdom
2026,536554,84670,NaN,23,2010-12-01 14:35:00,0.00,NaN,United Kingdom
2406,536589,21777,NaN,-10,2010-12-01 16:50:00,0.00,NaN,United Kingdom


In [54]:
# Count exact duplicate rows
duplicate_count = df.duplicated().sum()

print("Exact duplicate rows:", duplicate_count)

Exact duplicate rows: 5268


In [55]:
duplicates = df[df.duplicated(keep=False)]

display(duplicates.head(20))

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
485,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,"17,908.00",United Kingdom
489,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,"17,908.00",United Kingdom
494,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,"17,908.00",United Kingdom
517,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,"17,908.00",United Kingdom
521,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,"17,908.00",United Kingdom
527,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,"17,908.00",United Kingdom
537,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,"17,908.00",United Kingdom
539,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,"17,908.00",United Kingdom
548,536412,22327,ROUND SNACK BOXES SET OF 4 SKULLS,1,2010-12-01 11:49:00,2.95,"17,920.00",United Kingdom
555,536412,22327,ROUND SNACK BOXES SET OF 4 SKULLS,1,2010-12-01 11:49:00,2.95,"17,920.00",United Kingdom


In [56]:
print(
    "Unique invoices:",
    df["InvoiceNo"].nunique()
)

print(
    "Rows with repeated invoice numbers:",
    df["InvoiceNo"].duplicated().sum()
)

Unique invoices: 25900
Rows with repeated invoice numbers: 516009


In [57]:
print("Minimum quantity:", df["Quantity"].min())
print("Maximum quantity:", df["Quantity"].max())

print("Negative quantities:", (df["Quantity"] < 0).sum())
print("Zero quantities:", (df["Quantity"] == 0).sum())

Minimum quantity: -80995
Maximum quantity: 80995
Negative quantities: 10624
Zero quantities: 0


In [58]:
display(
    df[df["Quantity"] < 0]
    [["InvoiceNo", "Description",
      "Quantity", "InvoiceDate", "Country"]]
    .head(20)
)

,InvoiceNo,Description,Quantity,InvoiceDate,Country
141,C536379,Discount,-1,2010-12-01 09:41:00,United Kingdom
154,C536383,SET OF 3 COLOURED FLYING DUCKS,-1,2010-12-01 09:49:00,United Kingdom
235,C536391,PLASTERS IN TIN CIRCUS PARADE,-12,2010-12-01 10:24:00,United Kingdom
236,C536391,PACK OF 12 PINK PAISLEY TISSUES,-24,2010-12-01 10:24:00,United Kingdom
237,C536391,PACK OF 12 BLUE PAISLEY TISSUES,-24,2010-12-01 10:24:00,United Kingdom
238,C536391,PACK OF 12 RED RETROSPOT TISSUES,-24,2010-12-01 10:24:00,United Kingdom
239,C536391,CHICK GREY HOT WATER BOTTLE,-12,2010-12-01 10:24:00,United Kingdom
240,C536391,PLASTERS IN TIN VINTAGE PAISLEY,-12,2010-12-01 10:24:00,United Kingdom
241,C536391,PLASTERS IN TIN SKULLS,-24,2010-12-01 10:24:00,United Kingdom
939,C536506,JAM MAKING SET WITH JARS,-6,2010-12-01 12:38:00,United Kingdom


In [59]:
display(
    df.nlargest(10, "Quantity")[
        ["InvoiceNo", "Description",
         "Quantity", "UnitPrice"]
    ]
)

,InvoiceNo,Description,Quantity,UnitPrice
540421,581483,"PAPER CRAFT , LITTLE BIRDIE",80995,2.08
61619,541431,MEDIUM CERAMIC TOP STORAGE JAR,74215,1.04
502122,578841,ASSTD DESIGN 3D PAPER STICKERS,12540,0.00
74614,542504,NaN,5568,0.00
421632,573008,WORLD WAR 2 GLIDERS ASSTD DESIGNS,4800,0.21
206121,554868,SMALL POPCORN HOLDER,4300,0.72
220843,556231,?,4000,0.00
97432,544612,EMPIRE DESIGN ROSETTE,3906,0.82
270885,560599,ESSENTIAL BALM 3.5g TIN IN ENVELOPE,3186,0.06
52711,540815,FAIRY CAKE FLANNEL ASSORTED COLOUR,3114,2.10


In [60]:
print("Minimum price:", df["UnitPrice"].min())
print("Maximum price:", df["UnitPrice"].max())

print("Zero prices:", (df["UnitPrice"] == 0).sum())
print("Negative prices:", (df["UnitPrice"] < 0).sum())

Minimum price: -11062.06
Maximum price: 38970.0
Zero prices: 2515
Negative prices: 2


In [61]:
display(
    df[df["UnitPrice"] <= 0]
    [["InvoiceNo", "Description",
      "Quantity", "UnitPrice"]]
    .head(20)
)

,InvoiceNo,Description,Quantity,UnitPrice
622,536414,NaN,56,0.00
1970,536545,NaN,1,0.00
1971,536546,NaN,1,0.00
1972,536547,NaN,1,0.00
1987,536549,NaN,1,0.00
1988,536550,NaN,1,0.00
2024,536552,NaN,1,0.00
2025,536553,NaN,3,0.00
2026,536554,NaN,23,0.00
2406,536589,NaN,-10,0.00


In [62]:
display(
    df.nlargest(10, "UnitPrice")[
        ["InvoiceNo", "Description",
         "Quantity", "UnitPrice"]
    ]
)

,InvoiceNo,Description,Quantity,UnitPrice
222681,C556445,Manual,-1,"38,970.00"
524602,C580605,AMAZON FEE,-1,"17,836.46"
43702,C540117,AMAZON FEE,-1,"16,888.02"
43703,C540118,AMAZON FEE,-1,"16,453.71"
15016,C537630,AMAZON FEE,-1,"13,541.33"
15017,537632,AMAZON FEE,1,"13,541.33"
16356,C537651,AMAZON FEE,-1,"13,541.33"
16232,C537644,AMAZON FEE,-1,"13,474.79"
524601,C580604,AMAZON FEE,-1,"11,586.50"
299982,A563185,Adjust bad debt,1,"11,062.06"


In [63]:
print("Earliest invoice date:",
      df["InvoiceDate"].min())

print("Latest invoice date:",
      df["InvoiceDate"].max())

print("Missing invoice dates:",
      df["InvoiceDate"].isna().sum())

Earliest invoice date: 2010-12-01 08:26:00
Latest invoice date: 2011-12-09 12:50:00
Missing invoice dates: 0


In [64]:
print("InvoiceDate data type:",
      df["InvoiceDate"].dtype)

InvoiceDate data type: datetime64[us]


In [65]:
cancelled_mask = (
    df["InvoiceNo"]
    .astype(str)
    .str.startswith("C")
)

cancelled = df[cancelled_mask]

print("Cancelled invoice rows:", len(cancelled))

display(cancelled.head(10))

Cancelled invoice rows: 9288


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,2010-12-01 09:41:00,27.50,"14,527.00",United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,2010-12-01 09:49:00,4.65,"15,311.00",United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,2010-12-01 10:24:00,1.65,"17,548.00",United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,"17,548.00",United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,"17,548.00",United Kingdom
238,C536391,21980,PACK OF 12 RED RETROSPOT TISSUES,-24,2010-12-01 10:24:00,0.29,"17,548.00",United Kingdom
239,C536391,21484,CHICK GREY HOT WATER BOTTLE,-12,2010-12-01 10:24:00,3.45,"17,548.00",United Kingdom
240,C536391,22557,PLASTERS IN TIN VINTAGE PAISLEY,-12,2010-12-01 10:24:00,1.65,"17,548.00",United Kingdom
241,C536391,22553,PLASTERS IN TIN SKULLS,-24,2010-12-01 10:24:00,1.65,"17,548.00",United Kingdom
939,C536506,22960,JAM MAKING SET WITH JARS,-6,2010-12-01 12:38:00,4.25,"17,897.00",United Kingdom


In [66]:
cancellation_summary = pd.DataFrame({
    "Transaction Type": [
        "Cancelled",
        "Not Cancelled"
    ],
    "Rows": [
        cancelled_mask.sum(),
        (~cancelled_mask).sum()
    ]
})

display(cancellation_summary)

,Transaction Type,Rows
0,Cancelled,9288
1,Not Cancelled,532621


In [67]:
display(
    df.assign(
        IsCancelled=cancelled_mask,
        QuantityType=np.select(
            [
                df["Quantity"] < 0,
                df["Quantity"] == 0
            ],
            [
                "Negative",
                "Zero"
            ],
            default="Positive"
        )
    ).groupby(
        ["IsCancelled", "QuantityType"]
    ).size().unstack(fill_value=0)
)

QuantityType,Negative,Positive
IsCancelled,,
False,1336,531285
True,9288,0


In [68]:
quality_summary = pd.DataFrame({
    "Issue": [
        "Missing values (all columns)",
        "Exact duplicate rows",
        "Missing CustomerID",
        "Missing Description",
        "Missing InvoiceDate",
        "Negative quantities",
        "Zero quantities",
        "Zero prices",
        "Negative prices",
        "Cancelled invoice rows"
    ],
    "Count": [
        int(df.isnull().sum().sum()),
        int(df.duplicated().sum()),
        int(df["CustomerID"].isna().sum()),
        int(df["Description"].isna().sum()),
        int(df["InvoiceDate"].isna().sum()),
        int((df["Quantity"] < 0).sum()),
        int((df["Quantity"] == 0).sum()),
        int((df["UnitPrice"] == 0).sum()),
        int((df["UnitPrice"] < 0).sum()),
        int(cancelled_mask.sum())
    ]
})

display(quality_summary)

,Issue,Count
0,Missing values (all columns),136534
1,Exact duplicate rows,5268
2,Missing CustomerID,135080
3,Missing Description,1454
4,Missing InvoiceDate,0
5,Negative quantities,10624
6,Zero quantities,0
7,Zero prices,2515
8,Negative prices,2
9,Cancelled invoice rows,9288


### 5.2 Data Quality Findings

The investigation identified several data quality characteristics in the Online Retail dataset.

1. **Missing values:** The investigation identified missing values in CustomerID and Description. The most affected column was CustomerID with 135080 missing records.

2. **Duplicate records:** The dataset contained 5268 exact duplicate rows. Repeated invoice numbers were also observed, but these do not necessarily indicate duplicate transactions.

3. **Unusual quantities:** The dataset contained 10624 negative quantities and 0 zero quantities. Negative quantities require investigation because they may represent product returns.

4. **Unusual prices:** The investigation identified 2515 zero-priced records and 2 negative-priced records. These will be considered separately during cleaning.

5. **Dates:** The dataset covers the period from 2010-12-01 08:26:00 to 2011-12-09 12:50:00. There were 0 missing invoice dates.

6. **Cancellations:** The investigation identified 9288 cancelled invoice rows, which require separate treatment in sales analysis.

These findings will guide the cleaning process. Only records with a clear reason for removal will be excluded, while meaningful business transactions will be retained where possible.
